### ISSUE TITLE
jit_ir::GuardInst doesn't need `expect`

### ISSUE THREAD

**Issue by ltratt** [ykjit/yk#1437]
AFAICS `expect` on guards is pointless: you can always turn `guard true, X` into `guard false, Y`. For example `guard true, eq` is equivalent to `guard false, ne` and so on. Getting rid of `expect` (presumably in trace_builder) would simplify later parts of the pipeline.

**Comment by vext01** [comment]
For context, I think it was placed there with the intention of allowing more rich guard expressions, like `guard x < 4` or `guard x % 3 == 0`, but it may turn out that there are better ways to do this after all.

**Comment by ltratt** [comment]
`<`-esque comparisons don't need `true`/`false` (e.g. `false <` is `true >`) but `%` and friends would need it. Is this something that can happen in LLVM IR?

**Comment by vext01** [comment]
In LLVM IR a modulo check like that would be a modulo instruction, then a comparison to compute an `i1`,  so it probably doesn't make sense to have such rich expressions in our IR.

**Comment by ltratt** [comment]
Right, so it can't happen in our situation then?

**Comment by vext01** [comment]
It can happen, but if we wanted to bundle all of those semantics together into a guard, you'd have to consider several instructions together during lowering.

**Comment by ltratt** [comment]
Let me rephrase the question: can `Predicate` ever include operations that can't be negated? For example, if we need `Predicate::Modulo`, then that (maybe?) can't be negated. But if we don't need that, we don't need to worry. [We already consider multiple instructions during lowering: that's a separate matter.]

**Comment by vext01** [comment]
> can Predicate ever include operations that can't be negated?

I don't believe so: https://llvm.org/doxygen/classllvm_1_1CmpInst.html#a2be3583dac92a031fa1458d4d992c78b

**Comment by ltratt** [comment]
I now believe we can't do this because, at least hypothetically, we can have IR where we use an `i1` and guard against it: flipping the comparison would be unsound.
